File: src/api/gmail.ts
import { google } from "googleapis";
import { auth } from "../auth/auth.ts";
import { addExpense } from "../actions/actions.ts";

const EXPENSE_SYNCED_LABEL = "EXPENSE_SYNCED";
const AIRBNB_SUBJECT_FIRST_PART = "Reservation confirmed for";

const gmail = google.gmail({ version: "v1", auth });

async function createLabel(): Promise<string> {
  try {
    const response = await gmail.users.labels.create({
      userId: "me",
      requestBody: {
        name: EXPENSE_SYNCED_LABEL,
        labelListVisibility: "labelShow",
        messageListVisibility: "show",
      },
    });
    console.log(`Label created: ${response.data.name}`);
    return response.data.id as string;
  } catch (err) {
    if (err.code === 409) {
      console.log("Label already exists.");
      const existingLabels = await gmail.users.labels.list({ userId: "me" });
      const label = existingLabels.data.labels?.find(
        (label) => label.name === EXPENSE_SYNCED_LABEL,
      )!;
      return label.id as string;
    } else {
      console.error("Error creating label:", err);
      throw err;
    }
  }
}

async function getMessages() {
  console.log("Searching emails for sync expense");

  const response = await gmail.users.messages.list({
    userId: "me",
    maxResults: 100,
    q: `from:automated@airbnb.com subject:"${AIRBNB_SUBJECT_FIRST_PART}" is:unread -label:${EXPENSE_SYNCED_LABEL}`,
  });

  return response.data.messages;
}

async function markMessageAsSynced(messageId: string, syncLabelId: string) {
  await gmail.users.messages.modify({
    userId: "me",
    id: messageId,
    requestBody: {
      addLabelIds: [syncLabelId],
    },
  });
  console.log(`Message: ${messageId} marked as ${EXPENSE_SYNCED_LABEL}`);
}

export async function syncEmailsWithExpenses() {
  const messages = await getMessages();

  if (messages) {
    // TODO: think to move this outside of the fn
    const syncLabelId = await createLabel();

    for (const { id } of messages) {
      const message = await gmail.users.messages.get({
        userId: "me",
        id: id as string,
      });

      if (!message.data.payload?.headers) continue;

      const subject = message.data.payload.headers.find(
        (header) => header.name === "Subject",
      );

      if (!subject || !subject.value) continue;

      // get the checkin date and paid amount from message body parts
      const decodedString = atob(
        (message.data.payload?.parts?.[1].body?.data ?? "")
          .replace(/-/g, "+")
          .replace(/_/g, "/"),
      );

      const checkinDate = decodedString.match(/"checkinDate":"([^"]+)"/)?.[1];
      // const checkoutDate = decodedString.match(/"checkoutDate":"([^"]+)"/)?.[1];

      let totalPaid = decodedString
        .match(/Total \(USD\).*?>\$(\d{1,3}(?:,\d{3})*\.\d{2})<\/h3>/)?.[1]
        .replace(/,/g, "");

      const amountPaid = decodedString
        .match(/Amount paid \(USD\).*?\$([\d.]+)/)?.[1]
        .replace(/,/g, "");

      const amount = totalPaid ?? amountPaid;

      const description = subject.value.replace(
        AIRBNB_SUBJECT_FIRST_PART,
        "rent",
      );

      try {
        console.log(
          `Trying to add expense for ${description} on ${checkinDate} with amount ${Number(amount)}`,
        );
        await addExpense(description, Number(amount), checkinDate);
        console.log(`Added expense`);

        markMessageAsSynced(id as string, syncLabelId);
      } catch (err) {
        console.error("Error adding expense:", err);
      }
    }
  } else {
    console.log("No emails found for sync expense");
  }
}

File: src/api/gmail.py
import base64
import json

from bs4 import BeautifulSoup
from googleapiclient.errors import HttpError

from ..actions import actions
from ..auth.auth import service

EXPENSE_SYNCED_LABEL = "EXPENSE_SYNCED"
AIRBNB_SUBJECT_FIRST_PART = "Reservation confirmed for"


def create_label(service):
    try:
        response = (
            service.users()
            .labels()
            .create(
                userId="me",
                body={
                    "name": EXPENSE_SYNCED_LABEL,
                    "labelListVisibility": "labelShow",
                    "messageListVisibility": "show",
                },
            )
            .execute()
        )
        print(f"Label created: ${response.get('name')}")
        return response.get("id")
    except HttpError as err:
        if err.resp.status == 409:
            print("Label already exists.")
            existing_labels = service.users().labels().list(userId="me").execute()
            labels = existing_labels.get("labels", [])

            label = next(
                (l for l in labels if l.get("name") == EXPENSE_SYNCED_LABEL), None
            )
            if label:
                return label["id"]
        else:
            print("Error creating label:", err)
            raise err


# https://thepythoncode.com/article/use-gmail-api-in-python
def search_messages(service, query):
    result = service.users().messages().list(userId="me", q=query).execute()
    messages = []
    if "messages" in result:
        messages.extend(result["messages"])
    while "nextPageToken" in result:
        page_token = result["nextPageToken"]
        result = (
            service.users()
            .messages()
            .list(userId="me", q=query, pageToken=page_token)
            .execute()
        )
        if "messages" in result:
            messages.extend(result["messages"])
    return messages


def mark_message_as_synced(messages_to_mark, label_id):
    return (
        service.users()
        .messages()
        .batchModify(
            userId="me",
            body={
                "ids": [msg["id"] for msg in messages_to_mark],
                "addLabelIds": [label_id],
            },
        )
        .execute()
    )


def sync_email_with_expenses() -> None:
    query = f"from:automated@airbnb.com subject:'{AIRBNB_SUBJECT_FIRST_PART}' is:unread -label:{EXPENSE_SYNCED_LABEL}"
    results = search_messages(
        service,
        query,
    )
    if len(results) == 0:
        print("No emails found for sync expense")
        return

    sync_label_id = create_label(service)
    if sync_label_id is None:
        return

    messages_to_mark = []
    for message in results:
        print("message", message)
        msg = (
            service.users()
            .messages()
            .get(userId="me", id=message["id"], format="full")
            .execute()
        )
        payload = msg["payload"]

        headers = payload.get("headers")

        subject = None
        for header in headers:
            if header["name"] != "Subject":
                continue
            subject = header["value"]
            break

        if subject is None:
            continue

        print("headers name", subject)
        encoded_html = (
            payload.get("parts")[1]
            .get("body")
            .get("data")
            .replace("-", "+")
            .replace("_", "/")
        )

        html_string = base64.b64decode(encoded_html).decode("utf-8")
        soup = BeautifulSoup(html_string, "html.parser")
        script_tag = soup.find("script", {"data-testid": "siri-markup"})

        if script_tag is None:
            continue

        script_tag.get_text()
        json_text = script_tag.text.strip()
        data = json.loads(json_text)
        checkin_date = data.get("checkinDate")[:10]  # 2025-03-03
        # checkout_date = data.get("checkoutDate")[:10]

        amount_paid_element = soup.select(".right.heading3")
        if amount_paid_element[-1] is None:
            continue
        amount_str = amount_paid_element[-1].get_text()
        amount_clean_str = amount_str.replace("$", "").replace(",", "")
        amount = float(amount_clean_str)

        description = subject.replace(
            AIRBNB_SUBJECT_FIRST_PART,
            "rent",
        )

        print(
            f"Trying to add expense for {description} on {checkin_date} with amount {amount}"
        )
        actions.add_expense(description=description, amount=amount, date=checkin_date)
        messages_to_mark.append(message)

    len_messages = len(messages_to_mark)
    if len_messages > 0:
        print(f"Added {len_messages} expense{'s' if len_messages > 1 else ''}")
        mark_message_as_synced(messages_to_mark, sync_label_id)
        print("Marked messages as synced")
    else:
        print("Error adding expense")


def main():
    sync_email_with_expenses()


if __name__ == "__main__":
    main()
